### PR TITLE
feat: welcome page for external users with terms and conditions VOL-5664

### PR DESCRIPTION
**Pull request by ilindsay**
## Description

Welcome page now appears for external users to agree terms and conditions.

The following will need merging and dependencies bumped, before unit tests will pass:
[olcs-transfer](https://github.com/dvsa/olcs-transfer/pull/97)

Other related PRs:
[olcs-common](https://github.com/dvsa/olcs-common/pull/180)
[olcs-etl](https://github.com/dvsa/olcs-etl/pull/67)

Related issue: [VOL-5664](https://dvsa.atlassian.net/browse/VOL-5664)

### CODE REVIEW

File: app/api/module/Api/src/Entity/User/User.php [hunk 1]
```diff
@@ -765,4 +765,14 @@ public function canResetPassword(): bool
         //disabled users can't reset
         return !$this->isDisabled();
     }
+
+    public function agreeTermsAndConditions(): void
+    {
+        $this->termsAgreed = true;
```
[the remark this answers is not in the document]
Comment (ilindsay) [reply app/api/module/Api/src/Entity/User/User.php:771]: makes sense, I'll change it

File: app/api/module/Api/src/Entity/User/AbstractUser.php [hunk 2]
```diff
@@ -205,6 +205,15 @@ abstract class AbstractUser implements BundleSerializableInterface, JsonSerializ
...
     protected $translateToWelsh = 0;
 
+    /**
+     * Terms agreed
+     *
+     * @var boolean
+     *
+     * @ORM\Column(type="boolean", name="terms_agreed", nullable=false, options={"default": 0})
+     */
+    protected $termsAgreed = 0;
```
[the remark this answers is not in the document]
Comment (ilindsay) [reply app/api/module/Api/src/Entity/User/AbstractUser.php:215]: I don't know if I was in on that, if so I can't remember. But either way, what you suggest makes a lot of sense, and would be better all round. It'd require a few changes of course, including to ETL patch, so I'll check with Kab and see how bothered he his about it.

File: app/api/module/Api/src/Domain/CommandHandler/User/RegisterConsultantAndOperator.php [hunk 3]
```diff
@@ -32,10 +33,19 @@ public function handleCommand(CommandInterface $command)
...
+         *
+         * @var UserEntity $consultantUser
+         * @var RoleEntity $operatorTcRole
+         */
         $consultantUser = $this->getRepo()->fetchById($this->result->getId('user'));
         $operatorTcRole = $this->getRepo('Role')->fetchByRole(RoleEntity::ROLE_OPERATOR_TC);
         $consultantUser->setRoles(new ArrayCollection([$operatorTcRole]));
+
+        //the consultant has already accepted terms and conditions
+        $consultantUser->agreeTermsAndConditions();
```
[the remark this answers is not in the document]
Comment (ilindsay) [reply app/api/module/Api/src/Domain/CommandHandler/User/RegisterConsultantAndOperator.php:47]: I see what you mean - I'll have a look

File: app/api/module/Api/src/Entity/User/AbstractUser.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilindsay) [reply app/api/module/Api/src/Entity/User/AbstractUser.php:215]: Update: have spoken to Kab about this. Kab can only think of one time T&Cs have ever changed, going way back. So he doesn't consider it worth the work to change this over. Shame though, as we could have future proofed for little to no cost, had we done it earlier.

File: app/selfserve/module/Olcs/src/Controller/IndexController.php [hunk 4]
```diff
@@ -32,6 +32,10 @@ public function indexAction()
             throw new \Exception('Unable to retrieve identity');
         }
 
+        if (!$identity->getUserData()['termsAgreed']) {
```
[the remark this answers is not in the document]
Comment (ilindsay) [reply app/selfserve/module/Olcs/src/Controller/IndexController.php]: Agree. Switched to a listener.

File: app/api/module/Api/src/Domain/CommandHandler/User/RegisterConsultantAndOperator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (ilindsay) [reply app/api/module/Api/src/Domain/CommandHandler/User/RegisterConsultantAndOperator.php:51]: Added

**Review by fibble — APPROVED** [review]

**Review by sdh100shaun — DISMISSED** [review]

**Pull request merged**